File: rb/test/training_entity_test.rb
# Training entity test

require "minitest/autorun"
require "json"
require_relative "../AiCats_sdk"
require_relative "runner"

class TrainingEntityTest < Minitest::Test
  def test_create_instance
    testsdk = AiCatsSDK.test(nil, nil)
    ent = testsdk.Training(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = training_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "training." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set AICATS_TEST_TRAINING_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    training_ref01_ent = client.Training(nil)
    training_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.training"), "training_ref01"))

    training_ref01_data_result, err = training_ref01_ent.create(training_ref01_data, nil)
    assert_nil err
    training_ref01_data = Helpers.to_map(training_ref01_data_result)
    assert !training_ref01_data.nil?
    assert !training_ref01_data["id"].nil?

    # LIST
    training_ref01_match = {}

    training_ref01_list_result, err = training_ref01_ent.list(training_ref01_match, nil)
    assert_nil err
    assert training_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(training_ref01_list_result),
      { "id" => training_ref01_data["id"] })
    assert !Vs.isempty(found_item)

  end
end

def training_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "training", "TrainingTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = AiCatsSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["training01", "training02", "training03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["AICATS_TEST_TRAINING_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "AICATS_TEST_TRAINING_ENTID" => idmap,
    "AICATS_TEST_LIVE" => "FALSE",
    "AICATS_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["AICATS_TEST_TRAINING_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["AICATS_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = AiCatsSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["AICATS_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["AICATS_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/training_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/ai-cats-sdk/go"
	"github.com/voxgig-sdk/ai-cats-sdk/go/core"

	vs "github.com/voxgig-sdk/ai-cats-sdk/go/utility/struct"
)

func TestTrainingEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Training(nil)
		if ent == nil {
			t.Fatal("expected non-nil TrainingEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := trainingBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "training." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set AICATS_TEST_TRAINING_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		trainingRef01Ent := client.Training(nil)
		trainingRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "training"}, setup.data), "training_ref01"))

		trainingRef01DataResult, err := trainingRef01Ent.Create(trainingRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		trainingRef01Data = core.ToMapAny(trainingRef01DataResult)
		if trainingRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if trainingRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		trainingRef01Match := map[string]any{}

		trainingRef01ListResult, err := trainingRef01Ent.List(trainingRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		trainingRef01List, trainingRef01ListOk := trainingRef01ListResult.([]any)
		if !trainingRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", trainingRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(trainingRef01List), map[string]any{"id": trainingRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

	})
}

func trainingBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "training", "TrainingTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read training test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse training test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"training01", "training02", "training03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("AICATS_TEST_TRAINING_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"AICATS_TEST_TRAINING_ENTID": idmap,
		"AICATS_TEST_LIVE":      "FALSE",
		"AICATS_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["AICATS_TEST_TRAINING_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["AICATS_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewAiCatsSDK(core.ToMapAny(mergedOpts))
	}

	live := env["AICATS_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["AICATS_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
